src/cmdstan/command_helper.hpp

```diff
@@ -324,27 +358,87 @@ void get_fitted_params(const std::string &fname,
+ * @return a std vector of unconstrained parameter values
+ */
+std::vector<double> unconstrain_params(const stan::model::model_base &model,
+                                       const std::vector<double> &cparams) {
+  std::stringstream msg;
+  std::vector<std::string> param_names;
+  std::vector<std::vector<size_t>> param_dimss;
+  stan::services::get_model_parameters(model, param_names, param_dimss);
+  size_t num_uparams = model.num_params_r();
+  std::vector<double> uparams(num_uparams);
+  std::vector<int> dummy_params_i;
+  try {
+    stan::io::array_var_context context(param_names, cparams, param_dimss);
+    model.transform_inits(context, dummy_params_i, uparams, &msg);
+  } catch (const std::exception &e) {
+    std::stringstream msg2;
+    msg2 << e.what() << std::endl;
+    msg2 << "Bad or missing parameter values, cannot unconstrain.";
+    if (msg.str().length() > 0)
+      msg2 << "\n\t" << msg.str();
+    msg2 << std::endl;
+    throw std::invalid_argument(msg2.str());
+  }
+  return uparams;
+}
+
+/**
+ * Given an instantiated model and parsed StanCSV output file,
+ * apply model's "transform_init" to the fitted parameters.
+ * Returns a vector of vectors of parameters on the unconstrained scale.
+ * Throws an exception if the unconstraining transform failes.
+ *
+ * @param model instantiated model
+ * @param cparams vector of constrained param values
+ * @return a vector of std vectors of unconstrained parameter values
+ */
+std::vector<std::vector<double>> unconstrain_params_csv(
+    const stan::model::model_base &model, stan::io::stan_csv &fitted_params,
+    size_t &col_offset, size_t &num_rows, size_t &num_cols) {
+  std::stringstream msg;
+  std::vector<std::vector<double>> result;
+  for (size_t i = 0; i < num_rows; ++i) {
+    std::vector<double> uparams = unconstrain_params(
+        model, stan::math::to_array_1d(
+                   fitted_params.samples.block(i, col_offset, 1, num_cols)));
+    result.emplace_back(std::move(uparams));
+  }
+  return result;
+}
+
 /**
  * Parse a StanCSV output file created by the optimizer and return a vector
- * containing the estimates of the model parameters.
+ * containing the estimates of the model parameters on the unconstrained scale.
  * Throws an exception if it cannot parse the CSV file.
  *
- * @param fname name of file which exists and has read perms.
- * @param param_names vector of model constrained parameter names.
- * @param theta_hat Eigen vector for estimated parameters
+ * @param fname name of file which exists and has read perms
+ * @param model Stan model
+ * @return Eigen vector of unconstrained parameter estimates
  */
-void get_theta_hat_csv(const std::string &fname,
-                       const std::vector<std::string> &param_names,
-                       Eigen::VectorXd &theta_hat) {
+Eigen::VectorXd get_laplace_mode_csv(const std::string &fname,
+                                     const stan::model::model_base &model) {
```

Comment:
Line 475 is the culprit.

Reply:
I cannot trigger this segfault.
haven't changed the code.

Reply:
Well, UB isn’t guaranteed to segfault, but it’s pretty clear line 475 indexes out of bounds (in the currently pushed code) if `names.size() - cparam_names.size() < col_offset`.

The current if statements don’t catch this, since they only check that `names.size() >= cparam_names.size()`, so it won’t catch a model with 4 parameters being fed a CSV with lp_ and then 3 more columns after it, for example. This will then index out of bounds on `names`

Reply:
yes, managed to do this (hardcoded col offset to 10 - works like a charm - I just couldn't quite craft a CSV file that would make it blow up).  will tighten up the checks.